switchtec.c

```diff
@@ -150,6 +150,7 @@ static void mrpc_cmd_submit(struct switchtec_dev *stdev)
 	stdev->mrpc_busy = 1;
 	memcpy_toio(&stdev->mmio_mrpc->input_data,
 		    stuser->data, stuser->data_len);
+	wmb();
```

Comment:
This is wrong. Linux guarantees the ordering of IOs with respect to other IOs on the PCI bus and the CPU flushes the WC buffer on any load/store to a UC region (which is exactly the next command).  Even if this wasn't the case, a wmb() is a heavy tool to employ here.

Reply:
Oh wait scratch that... the next line is in the WC region. The correct way to handle this is a read from the device instead of a hard barrier. So issue an `ioread32(&stdev->mmio_mrpc->cmd)` and put a comment noting that it is there to flush the WC buffer.

Reply:
thx.
replace wmb() by ioread of reserved register in NTB db/msg register range
a. memrd tlp to this range of registers is processed by HW, with qucik response: ns level
b. out of this range, memrd tlp to the registers is processed by FW, with us or even ms level delay
c. dma mrpc feature is try to remove as many as possible of the ioread

Reply:
Reference from "DMA-API-HOWTO.txt"

.. important::

	     Consistent DMA memory does not preclude the usage of
	     proper memory barriers.  The CPU may reorder stores to
	     consistent memory just as it may normal memory.  Example:
	     if it is important for the device to see the first word
	     of a descriptor updated before the second, you must do
	     something like::

		desc->word0 = address;
		wmb();
		desc->word1 = DESC_VALID;

             in order to get correct behavior on all platforms.

	     Also, on some platforms your driver may need to flush CPU write
	     buffers in much the same way as it needs to flush write buffers
	     found in PCI bridges (such as by reading a register's value
	     after writing it).

Reply:
That quote is completely unrelated. This is talking about memory accesses that are also accessed by a DMA engine.

Linux and PCI ensure that IO writes are done in order. The only issue is that WC might delay the write until the next access. The correct way to flush the WC buffer is to issue a read from the device. That's what we did.

Logan

Reply:
@lsgunth 

Thx for comments.

But, i don't think Linux and PCI ensure that IO writes are done in order.
What I observed by PCIe analyzer's traces:
When WC feature(this patch) is not enabled for MRPC section: the MRPC input write is ordered
But when WC is applied: the reordering happend, we can find the payload for offset 0x80 is ahead of payload for 0x0.
So if there is no "read op" between MRPC input feeding and MRPC command firing, there will be reordering happened. That's not what we expected.

Also there is a patch #36 created for the same ordering issue.

Reply:
Reference: Documentation/io_ordering.txt

==============================================
Ordering I/O writes to memory-mapped addresses
==============================================

On some platforms, so-called memory-mapped I/O is weakly ordered.  On such
platforms, driver writers are responsible for ensuring that I/O writes to
memory-mapped addresses on their device arrive in the order intended.  This is
typically done by reading a 'safe' device or bridge register, causing the I/O
chipset to flush pending writes to the device before any reads are posted.

Reply:
Yes, you are correct. However, we did the right thing in patch #36, so I'm not sure what the issue is... We certainly shouldn't be adding wmb() calls anywhere. If there's another place that matters to add flush_wc_buf() then do so...

Reply:
@lsgunth

Thx for comments.

There is no issue.
I just want to get a clear picture of why wmb() is replaced by PCI read operation.


Regard,
Wesley

Reply:
wmb() is a hard memory barrier affecting all memory operations. It's very expensive.

Invoking a PCI read flushes the WC buffer so it accomplishes the same thing but a lot cheaper.

Reply:
@Isgunth

Thx for comments.

Regard,
Wesley